pkg/metrics/metrics.go

```diff
@@ -0,0 +1,246 @@
+		"Error can not write to GRPC",
+		TypeCounter,
+		"operational",
+	)
+	errCantNotWriteToKafka = defineMetric(
+		"err_cant_not_write_to_kafka",
+		"Error can not write to Kafka",
+		TypeCounter,
+		"operational",
+	)
+	errReadingRingBufferMap = defineMetric(
+		"err_reading_ring_buffer_map",
+		"Error reading ring buffer map",
+		TypeCounter,
+		"operational",
+	)
+	errCanNotDeleteFlowEntries = defineMetric(
+		"err_can_not_delete_flow_entries",
+		"Error can not delete flow entries",
+		TypeCounter,
+		"operational",
+	)
+)
+
+func (def *MetricDefinition) mapLabels(labels []string) prometheus.Labels {
+	if len(labels) != len(def.Labels) {
+		logrus.Errorf("Could not map labels, length differ in def %s [%v / %v]", def.Name, def.Labels, labels)
+	}
+	labelsMap := prometheus.Labels{}
+	for i, label := range labels {
+		labelsMap[def.Labels[i]] = label
+	}
+	return labelsMap
+}
+
+func verifyMetricType(def *MetricDefinition, t metricType) {
+	if def.Type != t {
+		logrus.Panicf("operational metric %q is of type %q but is being registered as %q", def.Name, def.Type, t)
+	}
+}
+
+type Metrics struct {
+	Settings *Settings
+}
+
+func NewMetrics(settings *Settings) *Metrics {
+	return &Metrics{Settings: settings}
+}
+
+// register will register against the default registry. May panic or not depending on settings
+func (m *Metrics) register(c prometheus.Collector, name string) {
+	err := prometheus.DefaultRegisterer.Register(c)
+	if err != nil {
+		if errors.As(err, &prometheus.AlreadyRegisteredError{}) {
+			logrus.Warningf("metrics registration error [%s]: %v", name, err)
+		} else if m.Settings.NoPanic {
+			logrus.Errorf("metrics registration error [%s]: %v", name, err)
+		} else {
+			logrus.Panicf("metrics registration error [%s]: %v", name, err)
+		}
+	}
+}
+
+func (m *Metrics) NewCounter(def *MetricDefinition, labels ...string) prometheus.Counter {
+	verifyMetricType(def, TypeCounter)
+	fullName := m.Settings.Prefix + def.Name
+	c := prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        fullName,
+		Help:        def.Help,
+		ConstLabels: def.mapLabels(labels),
+	})
+	m.register(c, fullName)
+	return c
+}
+
+func (m *Metrics) NewGauge(def *MetricDefinition, labels ...string) prometheus.Gauge {
+	verifyMetricType(def, TypeGauge)
+	fullName := m.Settings.Prefix + def.Name
+	c := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name:        fullName,
+		Help:        def.Help,
+		ConstLabels: def.mapLabels(labels),
+	})
+	m.register(c, fullName)
+	return c
+}
+
+func (m *Metrics) NewHistogram(def *MetricDefinition, buckets []float64, labels ...string) prometheus.Histogram {
+	verifyMetricType(def, TypeHistogram)
+	fullName := m.Settings.Prefix + def.Name
+	c := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:        fullName,
+		Help:        def.Help,
+		Buckets:     buckets,
+		ConstLabels: def.mapLabels(labels),
+	})
+	m.register(c, fullName)
+	return c
+}
+
+func (m *Metrics) CreateHashMapCounter(stage string) prometheus.Counter {
```

Comment:
In all the helper functions here, the "stage" param I think can be removed, that's something more relevant to FLP (because in FLP we configure custom pipelines, each with a stage name, so it was relevant to tie metrics to their stage)